packages/vite/src/node/plugins/resolve.ts

```diff
@@ -467,6 +416,66 @@ export default new Proxy({}, {
   }
 }
 
+function resolveSubpathImports(
+  id: string,
+  importer: string | undefined,
+  options: InternalResolveOptions,
+  targetWeb: boolean,
+) {
+  if (!importer || !id.startsWith(subpathImportsPrefix)) return
+  const basedir = path.dirname(importer)
+  const pkgData = findNearestPackageData(basedir, options.packageCache)
+  if (!pkgData) return
+
+  let importsPath = resolveExportsOrImports(
+    pkgData.data,
+    id,
+    options,
+    targetWeb,
+    'imports',
+  )
+
+  if (importsPath?.[0] === '.') {
+    importsPath = path.relative(basedir, path.join(pkgData.dir, importsPath))
+
+    if (importsPath[0] !== '.') {
+      importsPath = `./${importsPath}`
+    }
+  }
+
+  return importsPath
+}
+
+function ensureVersionQuery(
+  resolved: string,
+  id: string,
+  options: InternalResolveOptions,
+  depsOptimizer?: DepsOptimizer,
+): string {
+  if (
+    !options.isBuild &&
+    !options.scan &&
+    depsOptimizer &&
+    !(resolved === normalizedClientEntry || resolved === normalizedEnvEntry)
+  ) {
+    // Ensure that direct imports of node_modules have the same version query
+    // as if they would have been imported through a bare import
+    // Use the original id to do the check as the resolved id may be the real
+    // file path after symlinks resolution
+    const isNodeModule =
+      nodeModulesInPathRE.test(normalizePath(id)) ||
+      nodeModulesInPathRE.test(normalizePath(resolved))
```

Comment:
Maybe we can improve this part too. We're doing simple `node_modules` includes check everywhere, not sure why we special case a regex check here. We can simplify this to: `id.includes('node_modules') || resolved.includes('node_modules')`

Reply:
Ah it looks like I asked the same thing last year too 🙈 https://github.com/vitejs/vite/pull/9848/files#r956554867

Reply:
haha, and I'll say the same, but this time let's do it in another PR for real 😅

Reply:
Checking `node_modules` alone has worked well so far for us, I don't think we need to be stricter though